### PR TITLE
Preemptive interruption with Server Schedules

### PR DESCRIPTION
**Pull request by geraintpalmer**
This PR enables users to input a preemptive option with a server schedule, that interrupts services and resamples their service time after interruption. A virtual queue 'interrupted_individuals' keeps track of those interrupted individuals, ensuring they are re-served before new waiting customers.

### CODE REVIEW

**Comment by geraintpalmer** [comment]
@drvinceknight @Nikoleta-v3 Could you have a look over this before I merge please? Have I missed anything?

File: ciw/node.py [hunk 1]
```diff
@@ -144,7 +150,13 @@ def begin_service_if_possible_release(self, current_time):
         """
         if self.free_server() and self.c != float('Inf'):
             srvr = self.find_free_server()
-            if len([i for i in self.all_individuals if not i.server]) > 0:
+            if len(self.interrupted_individuals) > 0:
```
Comment (drvinceknight) [inline ciw/node.py]: This looks like its the same code as lines 133-139, could it be wrapped in a function/method?

Comment (geraintpalmer) [inline ciw/node.py]: Yes you're right, thanks! Will sort that now. Ta.

**Comment by geraintpalmer** [comment]
Happy to merge @drvinceknight ?

**Comment by drvinceknight** [comment]
> Happy to merge @drvinceknight ?

I'm running around. Did you throw in those two test cases we had up on the board/wall?

**Comment by geraintpalmer** [comment]
> Happy to merge @drvinceknight ?
> I'm running around. Did you throw in those two test cases we had up on the board/wall?

With slight alterations yes.

The only feature not implemented is the difference between Server and Till, i.e. queue capacity changing as servers go on and off duty. I think this should be another option (like preempt)

**Comment by drvinceknight** [comment]
> With slight alterations yes.

Cool, the only thing I see is perhaps missing is docs for this?

**Comment by geraintpalmer** [comment]
> Cool, the only thing I see is perhaps missing is docs for this?

Yep on purpose, as docs have changed in master. Avoiding conflicts. Will update docs as soon as merged. Thanks! Will merge now

**Pull request merged**